validator/src/state/merkle.rs

```diff
@@ -112,6 +115,65 @@ impl MerkleDatabase {
         })
     }
 
+    /// Prunes nodes that are no longer needed under a given state root
+    /// Returns a list of addresses that were deleted
+    pub fn prune(db: &LmdbDatabase, merkle_root: &str) -> Result<Vec<String>, MerkleDatabaseError> {
+        let root_bytes = ::hex::decode(merkle_root).map_err(|_| {
+            MerkleDatabaseError::InvalidHash(format!("{} is not a valid hash", merkle_root))
+        })?;
+        let mut db_writer = db.writer()?;
+        let change_log = get_change_log(&db_writer, &root_bytes)?;
+
+        if change_log.is_none() {
+            // There's no change log for this entry
+            return Ok(vec![]);
+        }
+
+        let change_log = change_log.unwrap();
+        let removed_addresses = if change_log.get_successors().len() > 1 {
+            // Currently, we don't clean up a parent with multiple successors
+            vec![]
+        } else if change_log.get_successors().len() == 0 {
+            // deleting the tip of a trie lineage
+            for hash in change_log.get_additions() {
+                let hash_hex = ::hex::encode(hash);
+                delete_ignore_missing(&mut db_writer, hash_hex.as_bytes())?
+            }
+
+            db_writer.index_delete(CHANGE_LOG_INDEX, &root_bytes)?;
+            let parent_root_bytes = &change_log.get_parent();
+
+            if let Some(ref mut parent_change_log) =
+                get_change_log(&db_writer, parent_root_bytes)?.as_mut()
+            {
+                let successors = parent_change_log.take_successors();
+                let new_successors = successors
+                    .into_iter()
+                    .filter(|successor| root_bytes != successor.get_successor())
+                    .collect::<Vec<_>>();
+                parent_change_log.set_successors(protobuf::RepeatedField::from_vec(new_successors));
```

Comment:
This might be doing some extra allocations that we may want to optimize in the future.

Reply:
Currently, `RepeatedField` from the protobuf library doesn't give us a `retain` method (like `Vec`), which would be the equivalent operation without extra allocations.

This isn't a high-frequency operation (most likely once per block), though, so I'm not sure it's going to be a big optimization point.